### PR TITLE
Cython

### PR DESCRIPTION
**Pull request by sfgraves**
This is a cython version of starlink-pyndf. I believe it is compatible with the previous version. The test suite has also been expanded to cover more of the basic NDF libraries functions, and to use a more complicated example data file.

### CODE REVIEW

File: test/test_hds.py [hunk 1]
```diff
@@ -25,31 +25,30 @@
...
+    'BOOL_ARRAY': ('_LOGICAL', [1,2,3], [[[ True, False,  True],[False,  False, False]]]),
+    'BOOL_ARRAY2': ('_LOGICAL', [1,2,3], [[[ False, True,  True],[True,  True, True]]])
 }
 
 keys = list(attributes_to_write.keys())
 keys.sort()
 
 def create_hds(testobj):
-    version = os.environ.get('HDS_VERSION', 4)
+    version = os.environ.get('HDS_VERSION',4)
```
Comment (timj) [inline test/test_hds.py]: Why is this space removed. It makes the file less PEP8 compatible.

File: setup.py [hunk 2]
```diff
@@ -142,11 +143,14 @@ def run(self):
             linked_libraries = []
             if ext.name=='starlink.hds':
 
+
```
Comment (timj) [inline setup.py]: Why was this line added?

File: test/test_hds.py [hunk 3]
```diff
@@ -67,25 +66,24 @@ def create_hds(testobj):
...
+                value = [i for i in np.asarray(value).flatten()][0]
         if comploc.type in ['_DOUBLE', '_REAL']:
             np.testing.assert_allclose(value,exp_value, verbose=True)
         else:
-
             np.testing.assert_equal(value, exp_value)
 
 
 
+
```
Comment (timj) [inline test/test_hds.py]: Too much whitespace 😄

File: test/test_simplendf.py [hunk 4]
```diff
@@ -23,12 +23,64 @@ def test_simpleread(self):
...
         self.assertEqual(indf.title, 'Test Data')
+        self.assertEqual(indf.xnumb, 1 )
+        self.assertEqual(indf.dim, [5,3])
+
+        # Xtension sutf
+        # xnumb (property), xname, xloc,  xstat, xdel,  xnew
+        self.assertEqual(indf.xname(0), 'FITS')
+        self.assertEqual(indf.xstat('FITS'), True)
+        self.assertEqual(indf.xstat('NONEXISTENT'), False)
+        self.assertIsInstance(type(indf.xloc('FITS', 'READ')), type(hds.HDSWrapperClass))
```
Comment (timj) [inline test/test_simplendf.py]: I'm confused by the `type()` calls here. `IsInstance` is meant to be checking instances and not classes.

File: test/test_simplendf.py [hunk 5]
```diff
@@ -23,12 +23,64 @@ def test_simpleread(self):
...
+
+
+        # AXSIS  stuff
```
Comment (timj) [inline test/test_simplendf.py]: Typo.

File: test/test_simplendf.py [hunk 6]
```diff
@@ -23,12 +23,64 @@ def test_simpleread(self):
...
+        # aform, astat, acget, anorm, amap,  aread,
+        self.assertEqual(indf.astat('LABEL', 0), True)
```
Comment (timj) [inline test/test_simplendf.py]: Please use `assertTrue` rather than `assertEqual` for booleans.

File: test/test_simplendf.py [hunk 7]
```diff
@@ -23,12 +23,64 @@ def test_simpleread(self):
...
+        self.assertEqual(history[0].application.split()[0], 'HISSET')
+        self.assertEqual(indf.type('DATA'), '_REAL')
+        self.assertSequenceEqual(indf.bound(), [[1,1], [5,3]])
+
+        data = indf.read('DATA')
+        data_shouldbe = (np.ones([5,3])*-3.4028235e+38).astype(np.float32)
+        self.assertSequenceEqual(data.tolist(), data_shouldbe.tolist())
+
+        #  cget, , new, map (and using it to change comp),
+        #  annul, , state
```
Comment (timj) [inline test/test_simplendf.py]: I don't really understand this comment or the `, ,`

File: test/test_simplendf.py [hunk 8]
```diff
@@ -23,12 +23,64 @@ def test_simpleread(self):
...
+        from starlink import Ast
+        self.assertIsInstance(indf.gtwcs(), Ast.FrameSet)
+
+        self.assertEqual(indf.state('LABEL'), True)
+        self.assertEqual(indf.state('QUALITY'), False)
+        with self.assertRaises(hds.StarlinkError):
+            indf.state('NONEXISTANT')
+
+        indf.annul()
```
Comment (timj) [inline test/test_simplendf.py]: Reminds me that this really needs a context manager so that you can do
```python
with ndf.open() as indf:
    stuff
```
and have it automatically close the file.

Comment (sfgraves) [reply test/test_simplendf.py]: Oh, thats a really good idea.

Comment (MalcolmCurrie) [reply test/test_simplendf.py]: The word nonexistant is non-existent.  (-:

**Comment by sfgraves** [comment]
I think I've addressed the main issues, so I'll merge this tomorrow unless anyone has further comments. (I haven't added a context manager yet, though thats definitely a good idea.)
